src/avatar_action.cpp

```diff
@@ -662,14 +662,27 @@ bool avatar_action::fire_check( avatar &you, const map &m, const targeting_data
         if( gun->get_gun_ups_drain() > 0 ) {
             const int ups_drain = gun->get_gun_ups_drain();
             const int adv_ups_drain = std::max( 1, ups_drain * 3 / 5 );
-
-            if( !( you.has_charges( "UPS_off", ups_drain ) ||
-                   you.has_charges( "adv_UPS_off", adv_ups_drain ) ||
-                   ( you.has_active_bionic( bionic_id( "bio_ups" ) ) && you.power_level >= ups_drain ) ) ) {
-                add_msg( m_info,
-                         _( "You need a UPS with at least %d charges or an advanced UPS with at least %d charges to fire that!" ),
-                         ups_drain, adv_ups_drain );
-                return false;
+            bool is_mech_weapon = false;
+            if( you.has_effect( effect_riding ) && g->u.mounted_creature ) {
+                auto mons = g->u.mounted_creature.get();
+                if( !mons->type->mech_weapon.empty() ) {
+                    is_mech_weapon = true;
+                }
+            }
+            if( !is_mech_weapon ) {
+                if( !( you.has_charges( "UPS_off", ups_drain ) ||
+                       you.has_charges( "adv_UPS_off", adv_ups_drain ) ||
+                       ( you.has_active_bionic( bionic_id( "bio_ups" ) ) && you.power_level >= ups_drain ) ) ) {
+                    add_msg( m_info,
+                             _( "You need a UPS with at least %d charges or an advanced UPS with at least %d charges to fire that!" ),
+                             ups_drain, adv_ups_drain );
+                    return false;
+                }
+            } else {
+                if( !you.has_charges( "UPS", ups_drain ) ) {
```

Comment:
So the mech weapon does not use advanced UPS or bionic UPS charges? Why not?

Reply:
Because the mechs weapon is wired directly into it's own battery.
THe use_charges( "UPS" ) will override and draw from the mechs battery.

Perhaps the player could jury-rig something to let the weapon take charges from their own bionic storage instead, but I can't see there being an easy access port for them to plug their own power supply in.